src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/S3EmitterUtils.scala

```diff
@@ -0,0 +1,197 @@
+    }
+  }
+
+
+  /**
+  * Terminate the application 
+  *
+  * Prevents shutdown hooks from running
+  */
+  private def forceShutdown(): Unit = {
+    log.error(s"Shutting down application as unable to connect to S3 for over $maxConnectionTime ms")
+    tracker foreach {
+      t =>
+        SnowplowTracking.trackApplicationShutdown(t)
+        sleep(5000)
+      }
+    Runtime.getRuntime.halt(1)
+  }
+
+
+  /**
+  * Returns an ISO valid timestamp
+  *
+  * @param tstamp The Timestamp to convert
+  * @return the formatted Timestamp
+  */
+  private def getTimestamp(tstamp: Long): String = {
+    val dt = new DateTime(tstamp)
+    TstampFormat.print(dt)
+  }
+
+  /**
+  * Sends records which fail deserialization or compression
+  * to NSQ with an error message
+  *
+  * @param records List of failed records to send to NSQ
+  */
+  def sendFailures(records: java.util.List[EmitterInput]): Unit = {
+    for (Failure(record) <- records.toList) {
+      log.warn(s"Record failed: $record.line")
+      log.info("Sending failed record to Kinesis")
+      val output = compact(render(
+        ("line" -> record.line) ~ 
+        ("errors" -> record.errors) ~
+        ("failure_tstamp" -> getTimestamp(System.currentTimeMillis()))
+      ))
+      badSink.store(output, Some("key"), false)
+    }
+  }
+
+  /**
+  * Keep attempting to send the data to S3 until it succeeds
+  *
+  * @return success status of sending to S3
+  */
+  def attemptEmit(namedStream: NamedStream, connectionAttemptStartTime: Long): Boolean = {
+
+    var attemptCount: Long = 1
+    while (true) {
+      if (attemptCount > 1 && System.currentTimeMillis() - connectionAttemptStartTime > maxConnectionTime) {
+        forceShutdown()
+      }
+
+      try {
+        val outputStream = namedStream.stream
+        val filename = namedStream.filename
+        val inputStream = new ByteArrayInputStream(outputStream.toByteArray)
+
+        val objMeta = new ObjectMetadata()
+        objMeta.setContentLength(outputStream.size.toLong)
+        client.putObject(bucket, filename, inputStream, objMeta)
+
+        return true
+      } catch {
+        // Retry on failure
+        case ase: AmazonServiceException => {
+          log.error("S3 could not process the request", ase)
+          tracker match {
+            case Some(t) => SnowplowTracking.sendFailureEvent(t, BackoffPeriod, connectionAttemptStartTime, attemptCount, ase.toString)
+            case None => None
+          }
+          attemptCount = attemptCount + 1
+          sleep(BackoffPeriod)
+        }
+        case NonFatal(e) => {
+          log.error("S3Emitter threw an unexpected exception", e)
+          tracker match {
+            case Some(t) => SnowplowTracking.sendFailureEvent(t, BackoffPeriod, connectionAttemptStartTime, attemptCount, e.toString)
+            case None => None
+          }
+          attemptCount = attemptCount + 1
+          sleep(BackoffPeriod)
+        }
+      }
+    }
+    false
+  }
+}
```

Comment:
This whole refactor needs to either have a dedicated issue or not be done at all.